README.md
# DIGIT-INTERFACE

[![License: CC BY-NC 4.0](https://img.shields.io/badge/License-CC%20BY--NC%204.0-lightgrey.svg)](LICENSE)
[![PyPI](https://img.shields.io/pypi/v/digit-interface)](https://pypi.org/project/digit-interface/)
<a href="https://digit.ml/">
<img height="20" src="/docs/digit-logo.svg" alt="DIGIT-logo" />
</a>

Python interface for the [DIGIT tactile sensor](https://digit.ml).

**For updates and discussions please join the #DIGIT channel at the [www.touch-sensing.org](https://www.touch-sensing.org/) community.**

## Installation

The preferred way of installation is through PyPi:

```bash
pip install digit-interface
```

Alternatively, you can manually clone the repository and install the package using:

```bash
git clone https://github.com/facebookresearch/digit-interface.git 
cd digit-interface
pip install -r requirements.txt
python setup.py install
```

If you cannot access the device by serial number on your system follow [adding DIGIT udev Rule](#adding-digit-udev-rule)

## Usage
The default connection method to the DIGIT tactile sensor is through the unique device serial number. The serial number
is found on the back of each DIGIT.
See [List all connected DIGIT's](#list-all-connected-digits) to find device serial numbers which are connected to the 
host.

Once you have the device serial number, reading data from the sensor should be as easy as
```python
from digit_interface import Digit
 
d = Digit("D12345") # Unique serial number
d.connect()
d.show_view()
d.disconnect()
```

Upon connection each DIGIT device initializes with a default stream resolution of ```VGA: 640x480``` at ```30fps```

#### Further Usage
##### List all connected DIGIT's:
To list all connected DIGIT's and display sensor information:
```python
from digit_interface import DigitHandler

digits = DigitHandler.list_digits()
```

##### Obtain a single frame:
```python
from digit_interface import Digit

d = Digit("D12345") # Unique serial number
d.connect()
frame = d.get_frame()
```

##### List supported stream formats:
Additional streams are supported, these streams vary in resolution and frames per second. 

To list the available stream formats:
```python
from digit_interface import Digit

print("Supported streams: \n {}".format(Digit.STREAMS))
```

##### Change resolution:
```
d.set_resolution(Digit.STREAMS["QVGA"])
```

##### Change FPS, 
Based on supported fps for each respective resolution. All streams support pre-defined resolutions which can 
be found in ```Digit.STREAMS```
```
d.set_fps(Digit.STREAMS["QVGA"]["fps"]["15fps"])
```

### Adding DIGIT udev Rule
Add your user to the ```plugdev``` group,

```
adduser username plugdev
```

Copy udev rule,

```
sudo cp ./udev/50-DIGIT.rules /lib/udev/rules.d/
```

Reload rules,

```
sudo udevadm control --reload
sudo udevadm trigger
```
 
Replug the DIGIT device into host.

## License
This code is licensed under CC-by-NC, as found in the [LICENSE](LICENSE) file.

## Citing
If you use this project in your research, please cite this [paper](https://arxiv.org/abs/2005.14679):

```BibTeX
@Article{Lambeta2020DIGIT,
  author  = {Lambeta, Mike and Chou, Po-Wei and Tian, Stephen and Yang, Brian and Maloon, Benjamin and Victoria Rose Most and Stroud, Dave and Santos, Raymond and Byagowi, Ahmad and Kammerer, Gregg and Jayaraman, Dinesh and Calandra, Roberto},
  title   = {{DIGIT}: A Novel Design for a Low-Cost Compact High-Resolution Tactile Sensor with Application to In-Hand Manipulation},
  journal = {IEEE Robotics and Automation Letters (RA-L)},
  year    = {2020},
  volume  = {5},
  number  = {3},
  pages   = {3838--3845},
  doi     = {10.1109/LRA.2020.2977257},
}
```


CHANGELOG.md
# Changelog

## 2021-04-27 [0.2.1]

- Changed location of Digit defaults
- Fixed error when printing device info when not connected

## 2021-04-27 [0.2]

- Added suport for individual control of RGB LED's through firmware version 2.00

## 2020-06-22 [0.1.4]

- Added PyPi support
- Added PyPi CI support
- Improved documentation

tests/test_import.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
import sys

import digit_interface


def test_main():
    if "digit_interface" not in sys.modules:
        print(f"Found digit_interface {digit_interface.__version__}")
        assert True


noxfile.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
import os

import nox

BASE = os.path.abspath(os.path.dirname(__file__))

DEFAULT_PYTHON_VERSIONS = ["3.7"]

LINT_SETUP_DEPS = ["black", "flake8", "flake8-copyright", "isort"]
DEPLOY_SETUP_DEPS = ["twine"]

VERBOSE = os.environ.get("VERBOSE", "0")
SILENT = VERBOSE == "0"

USING_CI = os.environ.get("USING_CI", False)
PYPI_USERNAME = os.environ.get("PYPI_USERNAME", None)
PYPI_PASSWORD = os.environ.get("PYPI_PASSWORD", None)


def _base_install(session):
    session.install("--upgrade", "setuptools", "pip", silent=SILENT)


def install_lint_deps(session):
    _base_install(session)
    session.install(*LINT_SETUP_DEPS, silent=SILENT)


def install_deploy_deps(session):
    _base_install(session)
    session.install(*DEPLOY_SETUP_DEPS, silent=SILENT)


def install_pytouch(session):
    session.chdir(BASE)
    session.run("pip", "install", "-e", ".")


@nox.session(python=DEFAULT_PYTHON_VERSIONS)
def lint(session):
    install_lint_deps(session)
    session.run("black", "--diff", "--check", ".", silent=SILENT)
    session.run(
        "isort",
        "--check",
        "--diff",
        ".",
        "--skip=.nox",
        silent=SILENT,
    )
    session.run("flake8", "--config", ".flake8")


@nox.session(python=DEFAULT_PYTHON_VERSIONS)
def tests(session):
    _base_install(session)
    install_pytouch(session)
    session.install("pytest")
    session.run("pytest", "tests")


@nox.session(python=DEFAULT_PYTHON_VERSIONS)
def build(session):
    _base_install(session)
    session.run("rm", "-rf", "dist", external=True)
    session.run("python", "setup.py", "sdist")


@nox.session(python=DEFAULT_PYTHON_VERSIONS)
def deploy(session):
    if not USING_CI:
        session.skip("Skipping deployment to PyPi.")
    install_deploy_deps(session)
    session.run(
        "twine",
        "upload",
        "dist/*",
        env={"TWINE_USERNAME": PYPI_USERNAME, "TWINE_PASSWORD": PYPI_PASSWORD},
    )


example/demo_digit.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.

# This source code is licensed under the license found in the LICENSE file in the root directory of this source tree.

import logging
import pprint
import time

import cv2

from digit_interface.digit import Digit
from digit_interface.digit_handler import DigitHandler

logging.basicConfig(level=logging.DEBUG)

# Print a list of connected DIGIT's
digits = DigitHandler.list_digits()
print("Connected DIGIT's to Host:")
pprint.pprint(digits)

# Connect to a Digit device with serial number with friendly name
digit = Digit("D12345", "Left Gripper")
digit.connect()

# Print device info
print(digit.info())

# Change LED illumination intensity
digit.set_intensity(Digit.LIGHTING_MIN)
time.sleep(1)
digit.set_intensity(Digit.LIGHTING_MAX)

# Change DIGIT resolution to QVGA
qvga_res = Digit.STREAMS["QVGA"]
digit.set_resolution(qvga_res)

# Change DIGIT FPS to 15fps
fps_30 = Digit.STREAMS["QVGA"]["fps"]["30fps"]
digit.set_fps(fps_30)

# Grab single frame from DIGIT
frame = digit.get_frame()
print(f"Frame WxH: {frame.shape[0]}{frame.shape[1]}")

# Display stream obtained from DIGIT
digit.show_view()

# Disconnect DIGIT stream
digit.disconnect()

# Find a Digit by serial number and connect manually
digit = DigitHandler.find_digit("D12345")
pprint.pprint(digit)
cap = cv2.VideoCapture(digit["dev_name"])
cap.release()


example/demo_rgb_intensity.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.

# This source code is licensed under the license found in the LICENSE file in the root directory of this source tree.

import logging
import pprint
import time

from digit_interface.digit import Digit
from digit_interface.digit_handler import DigitHandler

logging.basicConfig(level=logging.DEBUG)

# Print a list of connected DIGIT's
digits = DigitHandler.list_digits()
print("Connected DIGIT's to Host:")
pprint.pprint(digits)

# Connect to a Digit device with serial number with friendly name
digit = Digit("D12345", "Left Gripper")
digit.connect()

# Maximum value for each channel is 15
rgb_list = [(15, 0, 0), (0, 15, 0), (0, 0, 15)]

# Cycle through colors R, G, B
for rgb in rgb_list:
    digit.set_intensity_rgb(*rgb)
    time.sleep(1)

# Set all LEDs to same intensity
digit.set_intensity(15)

digit.disconnect()


digit_interface/digit_handler.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.

# This source code is licensed under the license found in the LICENSE file in the root directory of this source tree.

import logging
from typing import Dict, List, Optional

import pyudev

logger = logging.getLogger(__name__)


class DigitHandler:
    @staticmethod
    def _parse(digit_dev: Dict[str, str]) -> Dict[str, str]:
        digit_info = {
            "dev_name": digit_dev["DEVNAME"],
            "manufacturer": digit_dev["ID_VENDOR"],
            "model": digit_dev["ID_MODEL"],
            "revision": digit_dev["ID_REVISION"],
            "serial": digit_dev["ID_SERIAL_SHORT"],
        }
        return digit_info

    @staticmethod
    def list_digits() -> List[Dict[str, str]]:
        context = pyudev.Context()
        logger.debug("Finding udev devices with subsystem=video4linux, id_model=DIGIT")
        digits = context.list_devices(subsystem="video4linux", ID_MODEL="DIGIT")
        logger.debug("Following udev devices found: ")
        for device in digits:
            logger.debug(device)
        digits = [dict(DigitHandler._parse(_)) for _ in digits]
        if not digits:
            logger.debug("Could not find any udev devices matching parameters")
        return digits

    @staticmethod
    def find_digit(serial: str) -> Optional[Dict[str, str]]:
        digits = DigitHandler.list_digits()
        logger.debug(f"Searching for DIGIT with serial number {serial}")
        for digit in digits:
            if digit["serial"] == serial:
                return digit
        logger.error(f"No DIGIT with serial number {serial} found")
        return None


digit_interface/digit.py
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.

# This source code is licensed under the license found in the LICENSE file in the root directory of this source tree.

import logging
import typing

import cv2
import numpy as np

from digit_interface.digit_handler import DigitHandler

logger = logging.getLogger(__name__)


class DigitDefaults(object):
    STREAMS: typing.Dict = {
        # VGA resolution support 30 (default) and 15 fps
        "VGA": {
            "resolution": {"width": 640, "height": 480},
            "fps": {"30fps": 30, "15fps": 15},
        },
        # QVGA resolution support 60 (default) and 30 fps
        "QVGA": {
            "resolution": {"width": 320, "height": 240},
            "fps": {"60fps": 60, "30fps": 30},
        },
    }
    LIGHTING_MIN: int = 0
    LIGHTING_MAX: int = 15


class Digit(DigitDefaults):
    __LIGHTING_SCALER = 17

    def __init__(self, serial: str = None, name: str = None) -> None:
        """
        DIGIT Device class for a single DIGIT
        :param serial: DIGIT device serial
        :param name: Human friendly identifier name for the device
        """
        self.serial: str = serial
        self.name: str = name

        self.__dev: typing.Optional[cv2.VideoCapture] = None

        self.dev_name: str = ""
        self.manufacturer: str = ""
        self.model: str = ""
        self.revision: int = ""

        self.resolution: typing.Dict = {}
        self.fps: int = 0
        self.intensity: int = 0

        if self.serial is not None:
            logger.debug(f"Digit object constructed with serial {self.serial}")
            self.populate(serial)

    def connect(self) -> None:
        logger.info(f"{self.serial}:Connecting to DIGIT")
        self.__dev = cv2.VideoCapture(self.dev_name)
        if not self.__dev.isOpened():
            logger.error(
                f"Cannot open video capture device {self.serial} - {self.dev_name}"
            )
            raise Exception(f"Error opening video stream: {self.dev_name}")
        # set stream defaults, QVGA at 60 fps
        logger.info(
            f"{self.serial}:Setting stream defaults to QVGA, 60fps, maximum LED intensity."
        )
        logger.debug(f"Default stream to QVGA {self.STREAMS['QVGA']['resolution']}")
        self.set_resolution(self.STREAMS["QVGA"])
        logger.debug(f"Default stream with {self.STREAMS['QVGA']['fps']['60fps']} fps")
        self.set_fps(self.STREAMS["QVGA"]["fps"]["60fps"])
        logger.debug("Setting maximum LED illumination intensity")
        self.set_intensity(15)

    def set_resolution(self, resolution: typing.Dict) -> None:
        """
        Sets stream resolution based on supported streams in Digit.STREAMS
        :param resolution: QVGA or VGA from Digit.STREAMS
        :return: None
        """
        self.resolution = resolution["resolution"]
        width = self.resolution["width"]
        height = self.resolution["height"]
        logger.debug(f"{self.serial}:Stream resolution set to {height}w x {width}h")
        self.__dev.set(cv2.CAP_PROP_FRAME_WIDTH, width)
        self.__dev.set(cv2.CAP_PROP_FRAME_HEIGHT, height)

    def set_fps(self, fps: int) -> None:
        """
        Sets the stream fps, only valid values from Digit.STREAMS are accepted.
        This should typically be called after the resolution is set as the stream fps defaults to the
        highest fps
        :param fps: Stream FPS
        :return: None
        """
        self.fps = fps
        logger.debug(f"{self.serial}:Stream FPS set to {self.fps}")
        self.__dev.set(cv2.CAP_PROP_FPS, self.fps)

    def set_intensity(self, intensity: int) -> int:
        """
        Sets all LEDs to specific intensity, this is a global control.
        :param intensity: Value between 0 and 15 where 0 is all LEDs off and 15 all
        LEDS full intensity
        :return: Returns the set intensity
        """
        if self.revision < 200:
            # Deprecated version 1.01 (1b) is not supported
            intensity = int(intensity / self.__LIGHTING_SCALER)
            logger.warn(
                "You are using a previous version of the firmware "
                "which does not support independent RGB control, update your DIGIT firmware."
            )
        self.intensity = self.set_intensity_rgb(intensity, intensity, intensity)
        return self.intensity

    def set_intensity_rgb(
        self, intensity_r: int, intensity_g: int, intensity_b: int
    ) -> int:
        """
        Sets LEDs to specific intensity, per LED control
        Perimitted values are between 0 (off/dim) and 15 (full brightness)
        :param intensity_r: Red value
        :param intensity_g: Green value
        :param intensity_b: Blue value
        :return: Returns the set intensity
        """
        if not all(
            [x in range(0, 16) for x in (intensity_r, intensity_g, intensity_b)]
        ):
            raise ValueError("RGB values must be between 0 and 15.")
        intensity = (intensity_r << 8) | (intensity_g << 4) | intensity_b
        logger.debug(
            f"{self.serial}:LED intensity set to {intensity} (R: {intensity_r} G: {intensity_g} B: {intensity_b}"
        )
        self.intensity = intensity
        self.__dev.set(cv2.CAP_PROP_ZOOM, self.intensity)
        return self.intensity

    def get_frame(self, transpose: bool = False) -> np.ndarray:
        """
        Returns a single image frame for the device
        :param transpose: Show direct output from the image sensor, WxH instead of HxW
        :return: Image frame array
        """
        ret, frame = self.__dev.read()
        if not ret:
            logger.error(
                f"Cannot retrieve frame data from {self.serial}, is DIGIT device open?"
            )
            raise Exception(
                f"Unable to grab frame from {self.serial} - {self.dev_name}!"
            )
        if not transpose:
            frame = cv2.transpose(frame, frame)
            frame = cv2.flip(frame, 0)
        return frame

    def save_frame(self, path: str) -> np.ndarray:
        """
        Saves a single image frame to host
        :param path: Path and file name where the frame shall be saved to
        :return: None
        """
        frame = self.get_frame()
        logger.debug(f"Saving frame to {path}")
        cv2.imwrite(path, frame)
        return frame

    def get_diff(self, ref_frame: np.ndarray) -> np.ndarray:
        """
        Returns the difference between two frames
        :param ref_frame: Original frame
        :return: Frame difference
        """
        diff = self.get_frame() - ref_frame
        return diff

    def show_view(self, ref_frame: np.ndarray = None) -> None:
        """
        Creates OpenCV named window with live view of DIGIT device, ESC to close window
        :param ref_frame: Specify reference frame to show image difference
        :return: None
        """
        while True:
            frame = self.get_frame()
            if ref_frame is not None:
                frame = self.get_diff(ref_frame)
            cv2.imshow(f"Digit View {self.serial}", frame)
            if cv2.waitKey(1) == 27:
                break
        cv2.destroyAllWindows()

    def disconnect(self) -> None:
        logger.debug(f"{self.serial}:Closing DIGIT device")
        self.__dev.release()

    def info(self) -> str:
        """
        Returns DIGIT device info
        :return: String representation of DIGIT device
        """
        has_dev = self.__dev is not None
        is_connected = False
        if has_dev:
            is_connected = self.__dev.isOpened()
        info_string = (
            f"Name: {self.name} {self.dev_name}"
            f"\n\t- Model: {self.model}"
            f"\n\t- Revision: {self.revision}"
            f"\n\t- Connected?: {is_connected}"
        )
        if is_connected:
            info_string += (
                f"\nStream Info:"
                f"\n\t- Resolution: {self.resolution['width']} x {self.resolution['height']}"
                f"\n\t- FPS: {self.fps}"
                f"\n\t- LED Intensity: {self.intensity}"
            )
        return info_string

    def populate(self, serial: str) -> None:
        """
        Find the connected DIGIT based on the serial number and populate device parameters into the class
        :param serial: DIGIT serial number
        :return:
        """
        digit = DigitHandler.find_digit(serial)
        if digit is None:
            raise Exception(f"Cannot find DIGIT with serial {self.serial}")
        self.dev_name = digit["dev_name"]
        self.manufacturer = digit["manufacturer"]
        self.model = digit["model"]
        self.revision = int(digit["revision"])
        self.serial = digit["serial"]

    def __repr__(self) -> str:
        return f"Digit(serial={self.serial}, name={self.name})"
